internal/provider/aws/imds.go

```diff
@@ -0,0 +1,116 @@
+package aws
+
+import (
+	"fmt"
+	"io"
+	"net/http"
+	"sync"
+	"time"
+)
+
+const (
+	// Retrieve instance metadata for AWS EC2 instance
+	// https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-data-retrieval.html
+	instanceMetadataEndpoint = "http://169.254.169.254/latest/meta-data"
+
+	// IMDSv2 token related constants
+	tokenEndpoint      = "http://169.254.169.254/latest/api/token"
+	tokenTTLHeader     = "X-aws-ec2-metadata-token-ttl-seconds"
+	tokenRequestHeader = "X-aws-ec2-metadata-token"
+)
+
+// The VPC identifier
+// Automatically retrieved with GetVPCID function.
+// For run outside of the cluster, can be set through linker flag, e.g.
+// go build -ldflags "-X github.com/quortex/kubestatic/internal/provider/aws.vpcID=$VPC_ID" -a -o manager main.go
+var (
+	vpcID   string
+	vpcOnce sync.Once
+	vpcErr  error
+)
+
+// retrieveVPCID retrieves the VPC ID of the instance.
+// It first checks if the VPC ID is already cached. If not, it attempts to get an IMDSv2 token.
+// If the token retrieval fails, it falls back to IMDSv1. It then retrieves the MAC address
+// and uses it to get the VPC ID.
+func RetrieveVPCID(id string) (string, error) {
+	if id != "" {
+		vpcID = id
+	}
```

Comment:
I think since the function RetrieveVPCID is always called in the main, the id parameter and the singleton are not necessary. 

Instead if the user give us a vpcID we give that to the provider, and if not, the main could call RetrieveVPCID to retrieve it. This would have the benefit to remove both the singleton and the global variable.